File: client/src/lib/api/loyalty.ts
import { apiFetch } from "@/lib/api/client";

export type LoyaltyBadgeRead = {
  id: string;
  cinema_id: string | null;
  cinema_name: string | null;
  badge_key: string;
  title: string;
  description: string;
  earned_at: string;
};

export type RewardCouponRead = {
  id: string;
  cinema_id: string | null;
  status: "available" | "redeemed" | "expired";
  discount_percent: number;
  max_discount_cents: number;
  points_cost: number;
  expires_at: string;
  created_at: string;
};

export type VoucherOptionRead = {
  discount_percent: number;
  points_cost: number;
  max_discount_cents: number;
  available: boolean;
};

export type LoyaltyWalletRead = {
  points: number;
  level_name: string;
  next_level_name: string | null;
  points_to_next_level: number | null;
  badges: LoyaltyBadgeRead[];
  coupons: RewardCouponRead[];
  voucher_options: VoucherOptionRead[];
};

export function getMyLoyaltyWallet(token: string) {
  return apiFetch<LoyaltyWalletRead>("/loyalty/me", { token });
}

export function createRewardCoupon(discountPercent: number, token: string) {
  return apiFetch<RewardCouponRead>("/loyalty/coupons", {
    method: "POST",
    token,
    body: JSON.stringify({ discount_percent: discountPercent }),
  });
}

File: server/app/schemas/loyalty.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel


class LoyaltyBadgeRead(BaseModel):
    id: uuid.UUID
    cinema_id: uuid.UUID | None
    cinema_name: str | None = None
    badge_key: str
    title: str
    description: str
    earned_at: datetime


class RewardCouponRead(BaseModel):
    id: uuid.UUID
    cinema_id: uuid.UUID | None
    status: Literal["available", "redeemed", "expired"]
    discount_percent: int
    max_discount_cents: int
    points_cost: int
    expires_at: datetime
    created_at: datetime


class VoucherOptionRead(BaseModel):
    discount_percent: int
    points_cost: int
    max_discount_cents: int
    available: bool


class LoyaltyWalletRead(BaseModel):
    points: int
    level_name: str
    next_level_name: str | None
    points_to_next_level: int | None
    badges: list[LoyaltyBadgeRead]
    coupons: list[RewardCouponRead]
    voucher_options: list[VoucherOptionRead]


class CreateRewardCouponRequest(BaseModel):
    discount_percent: Literal[10, 20]

